### DISCUSSION TITLE
Mirror/flip transform on triangular mesh

### DISCUSSION THREAD

**Discussion by seblars** [isl-org/Open3D#6419] [General]
Hi,
I am trying to mirror transform the z-axis of a mesh. The transform works but I am getting weird lighting effect when I compute the vertex normals after the transform. It appears that the mesh is inside-out so the inside surface is reflective rather than the outside. I tried transforming the vertex normals without success. Is there another way? Example code below using version 0.17:

```
import open3d as o3d
import copy
import numpy as np

# mesh = o3d.geometry.TriangleMesh.create_coordinate_frame()

knot_mesh = o3d.data.KnotMesh()
mesh = o3d.io.read_triangle_mesh(knot_mesh.path)

I = np.eye(4) 
I[2,2] = -1
mesh_mirror = copy.deepcopy(mesh).transform(I)
mesh_mirror.compute_vertex_normals()
o3d.visualization.draw_geometries([mesh_mirror])
```

**Comment by benjaminum** [comment]
The mirroring causes the triangles orientation to flip. What you see is the inside of the mesh. You can try `mesh_mirror.orient_triangles()` or change the vertex index order of the triangles by manipulating `mesh_mirror.triangles`.

**Reply by seblars** [reply]
Thanks, reversing the order of ```mesh_mirror.triangles``` fixed the issue

**Reply by wmcnCG** [reply]
@seblars @benjaminum reversing the mesh triangles with mesh.triangles = mesh.triangles[::-1] gave me a random pattern. orient_triangles() did not work either. Am I missing something?

**Comment by cwreynolds** [comment]
Probably not directly to the questions of @seblars or @wmcnCG but just in case someone else working in c++ finds this Discussion, this is the code I use to “evert” (turn inside-out) a mesh. I use it for meshes meant to be seen from the inside:

```c++
// Flip orientation of each tri in a triangle mesh (destructively modifies).
void evertTriangleMesh(open3d::geometry::TriangleMesh& tri_mesh)
{
    for (auto& triangle : tri_mesh.triangles_)
    {
        std::reverse(std::begin(triangle), std::end(triangle));
    }
}
```

**Reply by wmcnCG** [reply]
Thank you. This python equivalent worked.
```
for i in range(len(mesh.triangles)):
        mesh.triangles[i] = mesh.triangles[i][::-1]
```

**Reply by seblars** [reply]
Agree - it is a useful workaround to make things transparent. If your mesh is large you can use a vectorised method such as:
```mesh.triangles = o3d.utility.Vector3iVector(np.flip(np.asarray(mesh.triangles), axis=1))```

**Reply by wmcnCG** [reply]
@seblars excellent!
